File: simulation/classes/Micromouse.cpp
/*
 * Implementation file for micromouse robot class
 * Author: Frederick Wachter
 */

#include <iostream>
#include <vector>

#include "Micromouse.h"
using namespace std;

// Move the robot forward (NOT USED DIRECTLY IN IMPLEMENTATION)
// Choose the only direction that is open
int Micromouse::chooseOpenDirection(void) {
	for (int i = 0; i < MAX_DIRECTIONS; i++) {
		if (robot.availableDirections[i] == 1) {
			direction = i;
			break;
		}
	}

	return direction;
}

// Choose a random direction to move based on input open directions
int Micromouse::chooseRandomDirection(const int input_availableDirections) {
	vector<int> indexes;

	for (int i = 0; i < MAX_DIRECTIONS; i++) {
		if (input_availableDirections[i] == 1) {
			indexes.push_back(i);
		}
	}

	direction = indexes.at(rand() % indexes.size());
	return direction;
}

// REturn the opposite direction of robot travel
int Micromouse::getOppositeDirection(void) {
	return (direction+2) % MAX_DIRECTIONS;
}

int Micromouse::getOppositeDirection(const int inputDirection) {
	return (inputDirection+2) % MAX_DIRECTIONS;
}

// Get total available directions from robot
int Micromouse::getTotalAvailableDirections(void) {
	int sum = 0;

	for (int i = 0; i < MAX_DIRECTIONS; i++) {
		sum += availableDirections[i];
	}

	return sum;
}

// Get total available directions from node
int Micromouse::getTotalAvailableDirections(const shared_ptr<Node> &node) {
	int sum = 0;

	for (int i = 0; i < MAX_DIRECTIONS; i++) {
		sum += node.availableDirections[i];
	}

	return sum;
}

bool Micromouse::isAtGoalLocation(void) {
	previousDirections.push_front(direction);
	if (previousDirections.size() == MAX_DIRECTIONS) {
		int sum = 0;
		for (int i = 0; i < (MAX_DIRECTIONS-1); i++) {
			sum += previousDirections[i];
			if (previousDirections[i] == robot.getOppositeDirection(previousDirections[i+1])) {
				return false;
			}
		}
		sum += previousDirections.back();
		if (sum == SUM_OF_DIRECTIONS) {
			return true;
		} else {
			return false;
		}
	}
}

void Micromouse::moveForward(void) {
	switch (direction) {
		case 0: location.y += 1;
		case 1: location.x += 1;
		case 2: location.y -= 1;
		case 3: location.x -= 1;
	}
}

// Set robot direction to opposite direction of the robot direction
int Micromouse::reverseDirection(void) {
	direction = (direction+2) % MAX_DIRECTIONS;

	return direction;
}

// Get the travel direction from start to goal
int Micromouse::setTravelDirection(const int xGoal, const int yGoal) {
	if ((location.x == xGoal) && (location.y != yGoal) {
		if (location.y > yGoal) {
			direction = 3;
		} else {
			direction = 1;
		}
	} else if ((location.y == yGoal) && (location.x != xGoal) {
		if (location.x > xGoal) {
			direction = 4;
		} else {
			direction = 2;
		}
	} else {
		cout << TRAVEL_DIRECTION_GOAL_ERROR << endl;
		return -1;
	}

	return direction;
}

File: simulation/classes/micromouse.m
%% ---- PROGRAM INFORMAITON ----
% PROGRAMMER: Frederick Wachter - wachterfreddy@gmail.com
% DATE CREATED: 2016-12-14
% PURPOSE: Micromouse functions class

% Please refer to the Wiki for instructions on how to use this class
% GITHUB WIKI: https://github.com/FWachter/Micromouse/wiki/MATLAB


classdef micromouse < handle
    
    properties
        location
        direction
        openDirections
        map
    end
    
    properties(GetAccess = private)
        goal
    end

%% CONSTRUCTOR METHOD
    methods
        
        function robot = micromouse(varargin)
        % EXAMPLE FUNCTION CALL: nodes = node(varargin)
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Construct the structure
            
            robot.location       = zeros(1, 2);
            robot.direction      = 0;
            robot.openDirections = zeros(1, 4);
            robot.map            = [];
            
            robot.goal.locations = zeros(5, 2);
        
        end
        
    end
    
%% PUBLIC METHODS
    
    methods
        
        function location = addCurrentLocation(robot, location)
        % EXAMPLE FUNCTION CALL: robot.addCurrentLocation(location)
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Adds provided location to array used to check for if the robot is in the goal position
            
            robot.goal.locations(2:5, :) = robot.goal.locations(1:4, :);
            robot.goal.locations(1, :)   = robot.location;
            
        end
        
        function goalLocationFound = isAtGoalLocation(robot)
        % EXAMPLE FUNCTION CALL: robot.isAtGoalLocation()
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Checks if the robot is at the goal position
            
            goalLocationFound = 0;
            if (sum(robot.goal.locations(1, :) == robot.goal.locations(5, :)) == 2) % if the robot found the goal
                directions = zeros(1, 4);
                for index = 1:4 % check the previous four directions traveled
                    directions(index) = robot.getTravelDirection(robot.goal.locations(index, :), robot.goal.locations(index+1, :));
                end
                
                for index = 1:3
                    if (robot.getOppositeDirection(directions(index)) == directions(index+1)) % if the set of locations was caused by a dead end not goal location
                        directions = zeros(1, 4);
                    end
                end
                
                if (length(unique(directions)) == 4) % if the robot went in a circle in the goal location
                    goalLocationFound = 1; 
                    fprintf('Solution found! Backtracking to remaining unsearched area\n');
                end
            end
            
        end
 
    end

    methods (Static = true)
        
        function oppositeDirection = getOppositeDirection(direction)
        % EXAMPLE FUNCTION CALL: robot.getOppositeDirection(direction)
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Get the opposite direction of a give direction
           
            oppositeDirection = mod(direction+2, 4);
            if (oppositeDirection == 0)
                oppositeDirection = 4;
            end
            
        end
        
        function direction = getTravelDirection(origin, goal)
        % EXAMPLE FUNCTION CALL: robot.getTravelDirection(origin, goal)
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Gets the neccessary direction of travel for going from an origin to a goal location
        % ASSUMPTION(S): A line can be drawn from the origin to the goal along one of the cartesian coordinates (same x or same y)
            
            if (goal(1) == origin(1))
                if (goal(2) > origin(2))
                    direction = 1;
                else
                    direction = 3;
                end
            else
                if (goal(1) > origin(1))
                    direction = 2;
                else
                    direction = 4;
                end
            end
            
        end
        
        function direction = chooseRandomDirection(openDirections)
        % EXAMPLE FUNCTION CALL: robot.chooseRandomDirection(openDirections)
        % PROGRAMMER: Frederick Wachter
        % DATE CREATED: 2016-12-14
        % PURPOSE: Choose a random direction from given open directions
            
            directions = find(openDirections == 1);
            randIndex = randi([1, length(directions)], 1);
            direction = directions(randIndex);
            
        end
        
    end 

%% PRIVATE METHODS

    methods (Access = private)
        
    end
    
end


